File: server/internal/protocol/terminal.go
package protocol

// TerminalOpenInput is the input for opening a new terminal session.
type TerminalOpenInput struct {
	ThreadId   ThreadId          `json:"threadId"`
	TerminalId string            `json:"terminalId"`
	Cwd        string            `json:"cwd"`
	Cols       *int              `json:"cols,omitempty"`
	Rows       *int              `json:"rows,omitempty"`
	Env        map[string]string `json:"env,omitempty"`
}

// TerminalWriteInput is the input for writing data to a terminal.
type TerminalWriteInput struct {
	ThreadId   ThreadId `json:"threadId"`
	TerminalId string   `json:"terminalId"`
	Data       string   `json:"data"`
}

// TerminalResizeInput is the input for resizing a terminal.
type TerminalResizeInput struct {
	ThreadId   ThreadId `json:"threadId"`
	TerminalId string   `json:"terminalId"`
	Cols       int      `json:"cols"`
	Rows       int      `json:"rows"`
}

// TerminalClearInput is the input for clearing a terminal.
type TerminalClearInput struct {
	ThreadId   ThreadId `json:"threadId"`
	TerminalId string   `json:"terminalId"`
}

// TerminalCloseInput is the input for closing a terminal.
type TerminalCloseInput struct {
	ThreadId      ThreadId `json:"threadId"`
	TerminalId    string   `json:"terminalId"`
	DeleteHistory *bool    `json:"deleteHistory,omitempty"`
}

// TerminalRestartInput is the input for restarting a terminal.
type TerminalRestartInput struct {
	ThreadId   ThreadId          `json:"threadId"`
	TerminalId string            `json:"terminalId"`
	Cwd        string            `json:"cwd"`
	Cols       int               `json:"cols"`
	Rows       int               `json:"rows"`
	Env        map[string]string `json:"env,omitempty"`
}

// TerminalSessionSnapshot is the state of a terminal session.
type TerminalSessionSnapshot struct {
	ThreadId   ThreadId `json:"threadId"`
	TerminalId string   `json:"terminalId"`
	Cwd        string   `json:"cwd"`
	Status     string   `json:"status"`
	Pid        *int     `json:"pid,omitempty"`
	History    string   `json:"history"`
	ExitCode   *int     `json:"exitCode,omitempty"`
	ExitSignal *int     `json:"exitSignal,omitempty"`
	UpdatedAt  string   `json:"updatedAt"`
}

// TerminalEvent is a push event for terminal state changes.
type TerminalEvent struct {
	ThreadId             ThreadId                 `json:"threadId"`
	TerminalId           string                   `json:"terminalId"`
	Type                 string                   `json:"type"`
	CreatedAt            string                   `json:"createdAt"`
	Data                 *string                  `json:"data,omitempty"`
	Snapshot             *TerminalSessionSnapshot `json:"snapshot,omitempty"`
	ExitCode             *int                     `json:"exitCode,omitempty"`
	ExitSignal           *int                     `json:"exitSignal,omitempty"`
	Message              *string                  `json:"message,omitempty"`
	HasRunningSubprocess *bool                    `json:"hasRunningSubprocess,omitempty"`
}

File: mobile/composeApp/src/commonMain/kotlin/io/j3mobile/protocol/Terminal.kt
package io.j3mobile.protocol

import kotlinx.serialization.Serializable

/** Input for opening a new terminal session. */
@Serializable
data class TerminalOpenInput(
    val threadId: ThreadId,
    val terminalId: String,
    val cwd: String,
    val cols: Int? = null,
    val rows: Int? = null,
    val env: Map<String, String>? = null,
)

/** Input for writing data to a terminal. */
@Serializable
data class TerminalWriteInput(
    val threadId: ThreadId,
    val terminalId: String,
    val data: String,
)

/** Input for resizing a terminal. */
@Serializable
data class TerminalResizeInput(
    val threadId: ThreadId,
    val terminalId: String,
    val cols: Int,
    val rows: Int,
)

/** Input for clearing a terminal. */
@Serializable
data class TerminalClearInput(
    val threadId: ThreadId,
    val terminalId: String,
)

/** Input for closing a terminal. */
@Serializable
data class TerminalCloseInput(
    val threadId: ThreadId,
    val terminalId: String,
    val deleteHistory: Boolean? = null,
)

/** Input for restarting a terminal. */
@Serializable
data class TerminalRestartInput(
    val threadId: ThreadId,
    val terminalId: String,
    val cwd: String,
    val cols: Int,
    val rows: Int,
    val env: Map<String, String>? = null,
)

/** State of a terminal session. */
@Serializable
data class TerminalSessionSnapshot(
    val threadId: ThreadId,
    val terminalId: String,
    val cwd: String,
    val status: String,
    val pid: Int? = null,
    val history: String,
    val exitCode: Int? = null,
    val exitSignal: Int? = null,
    val updatedAt: String,
)

/** Push event for terminal state changes. */
@Serializable
data class TerminalEvent(
    val threadId: ThreadId,
    val terminalId: String,
    val type: String,
    val createdAt: String,
    val data: String? = null,
    val snapshot: TerminalSessionSnapshot? = null,
    val exitCode: Int? = null,
    val exitSignal: Int? = null,
    val message: String? = null,
    val hasRunningSubprocess: Boolean? = null,
)
